collect_app/src/main/java/org/odk/collect/android/utilities/FormDownloader.java

```diff
@@ -0,0 +1,726 @@
+        try {
+            // get the xml file
+            // if we've downloaded a duplicate, this gives us the file
+            fileResult = downloadXform(fd.getFormName(), fd.getDownloadUrl());
+
+            if (fd.getManifestUrl() != null) {
+                // use a temporary media path until everything is ok.
+                tempMediaPath = new File(Collect.CACHE_PATH,
+                        String.valueOf(System.currentTimeMillis())).getAbsolutePath();
+                finalMediaPath = FileUtils.constructMediaPath(
+                        fileResult.getFile().getAbsolutePath());
+                String error = downloadManifestAndMediaFiles(tempMediaPath, finalMediaPath, fd,
+                        count, total);
+                if (error != null) {
+                    message += error;
+                }
+            } else {
+                Timber.i("No Manifest for: %s", fd.getFormName());
+            }
+        } catch (TaskCancelledException e) {
+            Timber.i(e.getMessage());
+            cleanUp(fileResult, e.file, tempMediaPath);
+
+            // do not download additional forms.
+            throw e;
+        } catch (Exception e) {
+            message += getExceptionMessage(e);
+        }
+
+        if (stateListener != null && stateListener.isTaskCanceled()) {
+            cleanUp(fileResult, null, tempMediaPath);
+            fileResult = null;
+        }
+
+        Map<String, String> parsedFields = null;
+        if (fileResult != null) {
+            try {
+                final long start = System.currentTimeMillis();
+                Timber.w("Parsing document %s", fileResult.file.getAbsolutePath());
+                parsedFields = FileUtils.parseXML(fileResult.file);
+                Timber.i("Parse finished in %.3f seconds.",
+                        (System.currentTimeMillis() - start) / 1000F);
+            } catch (RuntimeException e) {
+                message += e.getMessage();
+            }
+        }
+
+        boolean installed = false;
+
+        if ((stateListener == null || !stateListener.isTaskCanceled()) && message.isEmpty() && parsedFields != null) {
+            if (isSubmissionOk(parsedFields)) {
+                installEverything(tempMediaPath, fileResult, parsedFields);
+                installed = true;
+            } else {
+                message += Collect.getInstance().getString(R.string.xform_parse_error,
+                        fileResult.file.getName(), "submission url");
+            }
+        }
+        if (!installed) {
+            cleanUp(fileResult, null, tempMediaPath);
+        }
+        return message;
+    }
+
+    private boolean isSubmissionOk(Map<String, String> parsedFields) {
+        String submission = parsedFields.get(FileUtils.SUBMISSIONURI);
+        return submission == null || Validator.isUrlValid(submission);
+    }
+
+    private void installEverything(String tempMediaPath, FileResult fileResult, Map<String, String> parsedFields) {
```

Comment:
There maybe a better name for this method, it currently sounds a bit ambiguous and doesn't really tell the reader much about what it does.

Reply:
Maybe but it's the code I just moved from one class to another. Since this pr is pretty big itself we should do such a refactoring in a separate one.